### ISSUE TITLE
setHeader method may be set incorrectly

### ISSUE THREAD

**Issue by MikeMangum** [pillarjs/cookies#122]
The method used to set the header won't work for mocked requests:

`var setHeader = res.set ? http.OutgoingMessage.prototype.setHeader : res.setHeader`

Should probably be something along the lines of:

`var setHeader = res.setHeader ? res.setHeader : http.OutgoingMessage.prototype.setHeader`

**Comment by dougwilson** [comment]
The issue is that if it is reversed it will break what the workaround is for. There are some frameworks with a broken / different res.set , so this uses the original instead when it is there. But they have a (broken in the same way) res.setHeader as well.

**Comment by dougwilson** [comment]
This is the most recent comment that made the current implementation for reference: https://github.com/pillarjs/cookies/commit/e3b12470ed7c6169c582bcd206325e64411c283e

I will see if I can dig up the original issue / PR as well when I get back to a computer.

**Comment by dougwilson** [comment]
PR: https://github.com/pillarjs/cookies/pull/19

**Comment by MikeMangum** [comment]
My concern is that this is a workaround for a problem with a specific implementation of setHeader function but it is not targeted to that specific implementation.  
`var setHeader = !(res.__proto__.app === undefined) ? http.OutgoingMessage.prototype.setHeader : res.setHeader`

This would use the setHeader function of the passed in ServerResponse instance except in the specific case that it is express'
 implementation, which (if I understood the issue correctly) is the issue fixed in that PR.

**Comment by dougwilson** [comment]
I agree it is bad and I'm not sure i would have landed that pr if it were i, but it has been landed and is now part of the api. If you are ok with waiting a bit for me to get together a major release we can just remove it, otherwise just need a change that will not break folks .

**Comment by MikeMangum** [comment]
Sounds good.

**Comment by mrfitz42** [comment]
How goes the progress on this?  I'm working on unit tests for a system we are upgrading from node 4.4.7 and have found that this is causing an exception when using node-mocks-http to create res.

**Comment by si458** [comment]
im also having this problem,
im trying to use http2-express to enable http2.0 and it works but not with cookie-sessions
it appears cookie-sessions calls Cookies, then cookies.set
then it errors out saying `Cannot set properties of undefined (setting 'set-cookie')` from the line `setHeader.call`
and it appears that it doesnt like the http as im using http2 instead
so simply setting 
`var setHeader = res.setHeader ? res.setHeader : http.OutgoingMessage.prototype.setHeader`
does indeed fix my issue as it should be using the setHeader from the http2 and not http

**Comment by si458** [comment]
just circling back to this, Claude Code came up with this brilliant function and it works a treat!
just add it to your own project somewhere
```
if (args.http2 === true) {
  // Patch the 'cookies' module (used by cookie-session) for HTTP/2 compatibility.
  // cookies checks if res.set exists (Express adds it) and if so calls
  // http.OutgoingMessage.prototype.setHeader, which doesn't work on Http2ServerResponse
  // since it lacks the internal kOutHeaders symbol. We shadow res.set as undefined on
  // HTTP/2 responses during the call so cookies falls back to res.setHeader instead.
  const Cookies = require('cookies');
  const origCookiesSet = Cookies.prototype.set;
  Cookies.prototype.set = function(name, value, opts) {
      if (this.response.stream != null) {
          const res = this.response;
          const savedSetDesc = Object.getOwnPropertyDescriptor(res, 'set');
          Object.defineProperty(res, 'set', { value: undefined, configurable: true, writable: true });
          try { return origCookiesSet.call(this, name, value, opts); }
          finally { if (savedSetDesc) { Object.defineProperty(res, 'set', savedSetDesc); } else { delete res.set; } }
      }
      return origCookiesSet.call(this, name, value, opts);
  };
}
```
